File: Front/game_src/game_online/PowerUp.js
import { Config } from './Config.js'

export class PowerUp {
    constructor(scene, powerUpInfo) {
		this.scene = scene;
		this.powerUpInfo = powerUpInfo;
        this.arena = scene.arena;
		this.ball = scene.ball;

        // PowerUp object properties
        this.size = 0.05;
        this.geometry = new Config.THREE.SphereGeometry(this.size);
        this.material = new Config.THREE.MeshLambertMaterial({ 
            color: this.getClanColor(),
        });
        this.mesh = new Config.THREE.Mesh(this.geometry, this.material);
        this.initialZ = this.arena.arenaDepth / 2 + this.size;

        // true if PowerUp is illusion
		this.illusion = false;
    }

	getClanColor() {
        const clanColors = {
            'SCAVENGERS': 0x0000ff,    // Blue
            'RAIDERS': 0xff0000,    // Red
            'VERTEX': 0x00ff00   // Green
        };
        return clanColors[this.powerUpInfo.clan];
    }

    spawn() {
        this.mesh.position.set(this.powerUpInfo.x, this.powerUpInfo.y, this.initialZ);
        this.arena.mesh.add(this.mesh);
    }

    remove() {
        this.arena.mesh.remove(this.mesh);
    }

	checkExpired() {
		return Date.now() - this.spawnTime >= this.duration;
	}

    checkCollision() {
        if (!this.mesh)
			return false;
        
        const powerUpBounds = {
            minX: this.mesh.position.x - this.size/2,
            maxX: this.mesh.position.x + this.size/2,
            minY: this.mesh.position.y - this.size/2,
            maxY: this.mesh.position.y + this.size/2
        };

        const ballBounds = {
            minX: this.ball.mesh.position.x - this.ball.radius,
            maxX: this.ball.mesh.position.x + this.ball.radius,
            minY: this.ball.mesh.position.y - this.ball.radius,
            maxY: this.ball.mesh.position.y + this.ball.radius
        };

		const	isValidDirection = ((this.owner.side === 'left' && this.ball.velocityX > 0) ||
									(this.owner.side === 'right' && this.ball.velocityX < 0));
		if (!isValidDirection)
			return false;

        return !(powerUpBounds.minX > ballBounds.maxX || 
                powerUpBounds.maxX < ballBounds.minX || 
                powerUpBounds.minY > ballBounds.maxY || 
                powerUpBounds.maxY < ballBounds.minY);
    }

	applyPowerUp() {
		this.isActive = true;
		if (this.owner.clan == "SCAVENGERS"){
			this.sizePowerUp();
		}
		else if (this.owner.clan == "RAIDERS"){
			this.speedPowerUp();
		}
		else if (this.owner.clan == "VERTEX"){
			this.illusionPowerUp();
		}
	}

	speedPowerUp () {
		this.owner.paddleSpeed *= 1.5;
		setTimeout(() => {
			this.removeAbility();
        }, this.abilityDuration);
	}

	sizePowerUp () {
		this.owner.paddleHeight = 0.4;
		this.owner.updatePaddleGeometry(this.owner.mesh.position);

		setTimeout(() => {
			this.removeAbility();
        }, this.abilityDuration);
	}

	illusionPowerUp () {
		this.illusion = true;
		this.illusions = [];
		this.numIllusions = 2;

		for (let i = 0; i < this.numIllusions; i++) {
			const illusion = this.createIllusion();
			this.illusions.push(illusion);
			this.arena.mesh.add(illusion.mesh);
		}

		const updateIllusions = () => {
			if (this.scene.pause)
				return ;
			this.illusions.forEach(illusion => {
				illusion.mesh.position.x += illusion.velocityX;
				illusion.mesh.position.y += illusion.velocityY;

				// Bounce off vertical walls (top and bottom)
				if (Math.abs(illusion.mesh.position.y) >
							this.arena.arenaHeight / 2 - this.ball.radius) {
					illusion.velocityY = -illusion.velocityY;
					// Adjust position to prevent sticking
					illusion.mesh.position.y = Math.sign(illusion.mesh.position.y) *
						(this.arena.arenaHeight / 2 - this.ball.radius);
				}

				// Bounce off horizontal walls (left and right)
				if (Math.abs(illusion.mesh.position.x) >
							this.arena.arenaWidth / 2 - this.ball.radius) {
					illusion.velocityX = -illusion.velocityX;
					// Adjust position to prevent sticking
					illusion.mesh.position.x = Math.sign(illusion.mesh.position.x) *
						(this.arena.arenaWidth / 2 - this.ball.radius);
				}
			});
		};

        const intervalId = setInterval(updateIllusions, 16);
		this.illusionIntervalId = intervalId;

		// setTimeout(() => {
		// 	this.clearIllusion();
		// }, this.abilityDuration);
	}

	 createIllusion() {
		const geometry = new Config.THREE.SphereGeometry(this.ball.radius);
        const material = new Config.THREE.MeshLambertMaterial({
            color: this.ball.color,
            opacity: 0.5,
            transparent: true
        });
        const mesh = new Config.THREE.Mesh(geometry, material);
        
        // Random initial position near the real ball
        const initialX = this.ball.mesh.position.x + (Math.random() - 0.5) * 0.2;
        const initialY = this.ball.mesh.position.y + (Math.random() - 0.5) * 0.2;

		const xDirection = this.owner === 'left' ? 1 : -1;
        
        // Random velocity
        const speed = 0.02; // Adjust this to control illusion ball speed
        const angle = Math.random() * 2 * Math.PI;
        const velocityX = speed * Math.cos(angle) * xDirection;
        const velocityY = speed * Math.sin(angle);

        mesh.position.set(
            initialX,
            initialY,
            this.ball.mesh.position.z
        );

        return { 
            mesh, 
            velocityX, 
            velocityY
        };
    }

	clearIllusion () {
		if (this.illusionIntervalId){
			clearInterval(this.illusionIntervalId);
		}
		this.illusions.forEach(illusion => {
			this.arena.mesh.remove(illusion.mesh);
		});
		this.illusions = [];
	}

	removeAbility() {
		this.isActive = false;
		if (this.owner.clan == "SCAVENGERS"){
			this.owner.paddleHeight = this.originalSize;
			this.owner.updatePaddleGeometry(this.owner.mesh.position);
		}
		else if (this.owner.clan == "RAIDERS"){
			this.owner.paddleSpeed = this.originalSpeed;
		}
		else if (this.owner.clan == "VERTEX"){
				this.clearIllusion();
		}
	}
}

File: Back/backend/game/power_up.py
import random
import math
import time
import threading

class PowerUp:
    def __init__(self, arena, owner, ball):
        self.arena = arena
        self.owner = owner
        self.clan = self.owner.clan
        self.ball = ball

        self.is_active = False
        self.original_owner_size = owner.paddle_height
        self.original_owner_speed = owner.paddle_speed

        self.size = 0.05
        self.spawn_time = time.time()
        self.duration = 10
        self.activation_duration = 10

        self.x = 0
        self.y = 0

    def spawn(self):
        max_x = (self.arena.arena_width / 2) - (self.size * 2)
        max_y = (self.arena.arena_height / 2) - (self.size * 2)
        
        self.x = (random.random() * 2 - 1) * max_x
        self.x = (random.random() * 2 - 1) * max_x
        # self.x = 0.5
        # self.y = 0

    def check_expired(self):
        return time.time() - self.spawn_time >= self.duration

    def check_collision(self):
        """
        Check if the ball collides with the power-up.
    
        :return: Boolean indicating collision
        """
        # Check if the ball is moving in the correct direction
        is_valid_direction = (
            (self.owner.side == 'left' and self.ball.vx > 0) or
            (self.owner.side == 'right' and self.ball.vx < 0)
        )
        
        if not is_valid_direction:
            return False
        
        # Collision detection
        power_up_bounds = {
            'min_x': self.x - self.size/2,
            'max_x': self.x + self.size/2,
            'min_y': self.y - self.size/2,
            'max_y': self.y + self.size/2
        }
        
        ball_bounds = {
            'min_x': self.ball.x - self.ball.radius,
            'max_x': self.ball.x + self.ball.radius,
            'min_y': self.ball.y - self.ball.radius,
            'max_y': self.ball.y + self.ball.radius
        }
        
        return not (
            power_up_bounds['min_x'] > ball_bounds['max_x'] or 
            power_up_bounds['max_x'] < ball_bounds['min_x'] or 
            power_up_bounds['min_y'] > ball_bounds['max_y'] or 
            power_up_bounds['max_y'] < ball_bounds['min_y']
        )

    def apply_power_up(self):
        self.is_active = True

        if self.owner.clan == 'SCAVENGERS':
            self.size_power_up()
        elif self.owner.clan == 'RAIDERS':
            self.speed_power_up()
        elif self.owner.clan == 'VERTEX':
            self.illusion_power_up()

    def speed_power_up(self):
        self.owner.paddle_speed *= 1.5

        timer = threading.Timer(self.activation_duration, self.remove_ability)
        timer.start()

    def size_power_up(self):
        self.owner.paddle_height = 0.4

        timer = threading.Timer(self.activation_duration, self.remove_ability)
        timer.start()

    def illusion_power_up(self):
        self.owner.illusion = True

        timer = threading.Timer(self.activation_duration, self.remove_ability)
        timer.start()
        # self.illusions = []
        # num_illusions = 2
        #
        # for _ in range(num_illusions):
        #     illusion = self.houdini_phantomforge()
        #     self.illusions.append(illusion)

    # def houdini_phantomforge(self):

    def remove_ability(self):
        """
        Remove the power-up ability and reset paddle to original state.
        """
        self.is_active = False

        # self.owner.paddle_speed = self.original_owner_speed
        
        if self.owner.clan == "SCAVENGERS":
            self.owner.paddle_height = self.original_owner_size
        elif self.owner.clan == "RAIDERS":
            self.owner.paddle_speed = self.original_owner_speed
        elif self.owner.clan == "VERTEX":
            self.owner.illusion = False
